pynamodb/attributes.py

```diff
@@ -959,23 +965,43 @@ def deserialize(self, values):
         """
         deserialized_lst = []
         for v in values:
-            class_for_deserialize = self.element_type() if self.element_type else _get_class_for_deserialize(v)
-            attr_value = _get_value_for_deserialize(v)
-            deserialized_lst.append(class_for_deserialize.deserialize(attr_value))
+            attr_type, attr_value = next(iter(v.items()))
+            attr_class = self._get_deserialize_class(attr_type)
+            if self.element_type and attr_type == NULL:
+                # When attribute values serialize to "None" (e.g. empty sets) we store {"NULL": True} in DynamoDB.
+                attr_type = self.element_type.attr_type
+                attr_value = None
+            if self.element_type and self.element_type.attr_type != attr_type:
+                raise ValueError("Cannot deserialize elements of type: {}".format(attr_type))
```

Comment:
Can we add the expected `attr_type` to this message too? Just to help folks understand the issue more quickly

Reply:
updated with attribute class name